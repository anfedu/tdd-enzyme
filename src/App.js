import React from "react";
import { useColorMode, Box, Button, IconButton } from "@chakra-ui/react";
import { SunIcon, MoonIcon } from "@chakra-ui/icons";
import Lists from "./components/Lists";

function App() {
  const { colorMode, toggleColorMode } = useColorMode();
  const people = [
    { person: "Nuril" },
    { person: "Rizqi" },
    { person: "Imroatul" },
    { person: "Amalia" },
  ];

  return (
    <Box className="h-[100vh] flex justify-center items-center gap-x-5">
      <Button onClick={toggleColorMode}>
        {colorMode === "light" ? "dark" : "light"} mode
      </Button>
      <IconButton
        role="iconButton"
        onClick={toggleColorMode}
        icon={colorMode === "dark" ? <SunIcon /> : <MoonIcon />}
      />
      <div>
        {people.map((item, key) => (
          <div key={key}>
            <Lists role={`lists`} roleId={`list-${key}`} item={item} />
          </div>
        ))}
      </div>
    </Box>
  );
}

export default App;
