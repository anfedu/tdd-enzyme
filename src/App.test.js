import App from "./App";
import { shallow } from "enzyme";
import { useColorMode } from "@chakra-ui/react";

jest.mock("@chakra-ui/icons", () => {
  return {
    SunIcon: "SunIcon",
    MoonIcon: "MoonIcon",
  };
});
jest.mock("@chakra-ui/react", () => {
  return {
    Box: "Box",
    Button: "Button",
    IconButton: "IconButton",
    useColorMode: jest.fn(),
  };
});

describe("App", () => {
  let wrapper;
  let toggleColorModeMock = jest.fn();
  beforeEach(() => {
    // <-- must declare first -->
    useColorMode.mockReturnValue({
      colorMode: "dark",
      toggleColorMode: toggleColorModeMock,
    });

    // <-- must last declare -->
    wrapper = shallow(<App />);
  });

  // <-- snap and check changed file -->
  it("render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  // <-- check existing elements -->
  it("Lists component", () => {
    const listsElement = wrapper.find("[role='lists']").exists();
    expect(listsElement).toBe(true);
  });

  // <-- check props -->
  it("List component", () => {
    const listElement = wrapper.find("[roleId='list-0']").props();
    expect(listElement.item.person).toBe("Nuril");
  });

  it("When app on dark mode, component should render properly", () => {
    const iconProps = wrapper.find("[role='iconButton']").props();
    expect(iconProps.icon.type).toBe("SunIcon");

    iconProps.onClick();
    expect(toggleColorModeMock).toHaveBeenCalled();
  });

  // it("When app on light mode, component should render properly", () => {
  //   const iconProps = wrapper.find("[role='iconButton']").props();
  //   expect(iconProps.icon.type).toBe("MoonIcon");

  //   iconProps.onClick();
  //   expect(toggleColorModeMock).toHaveBeenCalled();
  // });
});
