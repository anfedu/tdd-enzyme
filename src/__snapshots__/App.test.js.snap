// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render 1`] = `
<Box
  className="h-[100vh] flex justify-center items-center gap-x-5"
>
  <Button
    onClick={[MockFunction]}
  >
    light
     mode
  </Button>
  <IconButton
    icon={<SunIcon />}
    onClick={[MockFunction]}
    role="iconButton"
  />
  <div>
    <div
      key="0"
    >
      <Lists
        item={
          Object {
            "person": "Nuril",
          }
        }
        role="lists"
        roleId="list-0"
      />
    </div>
    <div
      key="1"
    >
      <Lists
        item={
          Object {
            "person": "Rizqi",
          }
        }
        role="lists"
        roleId="list-1"
      />
    </div>
    <div
      key="2"
    >
      <Lists
        item={
          Object {
            "person": "Imroatul",
          }
        }
        role="lists"
        roleId="list-2"
      />
    </div>
    <div
      key="3"
    >
      <Lists
        item={
          Object {
            "person": "Amalia",
          }
        }
        role="lists"
        roleId="list-3"
      />
    </div>
  </div>
</Box>
`;
